utils/stall_free.hh

```diff
@@ -57,14 +59,176 @@ void merge_to_gently(std::list<T>& list1, const std::list<T>& list2, Compare com
     }
 }
 
-template<class T>
-seastar::future<> clear_gently(std::list<T>& list) {
-    return repeat([&list] () mutable {
-        if (list.empty()) {
-            return seastar::stop_iteration::yes;
-        }
-        list.pop_back();
-        return seastar::stop_iteration::no;
+// The clear_gently functions are meant for
+// gently destroying the contents of containers.
+// The containers can be re-used after clear_gently
+// or may be destroyed.  But unlike e.g. std::vector::clear(),
+// clear_gently will not necessarily keep the object allocation.
+
+template <typename T>
+concept HasClearGentlyMethod = requires (T x) {
+    { x.clear_gently() } -> std::same_as<future<>>;
+};
+
+template <typename T>
+concept SmartPointer = requires (T x) {
+    { x.get() } -> std::same_as<typename T::element_type*>;
+};
+
+template <typename T>
+concept SharedPointer = SmartPointer<T> && requires (T x) {
+    { x.use_count() } -> std::convertible_to<long>;
+};
+
+template <typename T>
+concept StringLike = requires (T x) {
+    std::is_same_v<typename T::traits_type, std::char_traits<typename T::value_type>>;
+};
+
+template <typename T>
+concept Iterable = requires (T x) {
+    { x.empty() } -> std::same_as<bool>;
+    { x.begin() } -> std::same_as<typename T::iterator>;
+    { x.end() } -> std::same_as<typename T::iterator>;
+};
+
+template <typename T>
+concept Sequence = Iterable<T> && requires (T x, size_t n) {
+    { x.back() } -> std::same_as<typename T::value_type&>;
+    { x.pop_back() } -> std::same_as<void>;
+};
+
+template <typename T>
+concept TriviallyClearableSequence =
+    Sequence<T> && std::is_trivially_destructible_v<typename T::value_type> && !HasClearGentlyMethod<typename T::value_type>;
+
+template <typename T>
+concept Container = Iterable<T> && requires (T x, typename T::iterator it) {
+    { x.erase(it) } -> std::same_as<typename T::iterator>;
+};
+
+template <typename T>
+concept MapLike = Container<T> && requires (T x) {
+    std::is_same_v<typename T::value_type, std::pair<typename T::key_type, typename T::mapped_type>>;
```

Comment:
Does this work? I think a map's value_type is std::pair<const Key, Value>.

Reply:
You're right about the value_type.
For some reason, this works when defined as a concept (but not otherwise)
See https://godbolt.org/z/Eo83Mrfc3

I'll add `const`

Reply:
Fixed in v7